atom/browser/api/atom_api_system_preferences_mac.mm

```diff
@@ -360,6 +361,61 @@ static bool FromV8(v8::Isolate* isolate,
   }
 }
 
+// whether the system has access to camera
+bool SystemPreferences::HasCameraAccess() {
+  return [[AtomAccessController sharedController] hasCameraAccess];
+}
+
+// whether the system has access to microphone
+bool SystemPreferences::HasMicrophoneAccess() {
+  return [[AtomAccessController sharedController] hasMicrophoneAccess];
+}
+
+// whether the system has access to both microphone and camera
+bool SystemPreferences::HasFullMediaAccess() {
+  return [[AtomAccessController sharedController] hasFullMediaAccess];
+}
+
+// ask for access to camera and/or microphone
+v8::Local<v8::Promise> SystemPreferences::AskForMediaAccess(
+    v8::Isolate* isolate,
+    mate::Arguments* args) {
+  scoped_refptr<util::Promise> promise = new util::Promise(isolate);
+  std::string media_type;
+
+  bool ask_again = false;
+  if (!args->GetNext(&ask_again)) {
+    promise->RejectWithErrorMessage("askAgain value is required");
+    return promise->GetHandle();
+  }
+
+  if (args->GetNext(&media_type)) {
+    if (media_type == "microphone")
+      [[AtomAccessController sharedController]
+          askForMicrophoneAccess:ask_again
+                      completion:^(BOOL granted) {
+                        promise->Resolve(granted == YES);
+                      }];
+    else if (media_type == "camera") {
+      [[AtomAccessController sharedController]
+          askForCameraAccess:ask_again
+                  completion:^(BOOL granted) {
+                    promise->Resolve(granted == YES);
+                  }];
+    } else {
+      promise->RejectWithErrorMessage(
+          "Invalid media type, use 'camera' or 'microphone'.");
+    }
+  } else {
+    [[AtomAccessController sharedController]
+        askForMediaAccess:ask_again
+               completion:^(BOOL granted) {
+                 promise->Resolve(granted == YES);
```

Comment:
If `granted` is already a `BOOL`, what't the point tot compare it to `YES`?

Reply:
objective-C BOOLs and c++ booleans don't 1:1 match up; `promise->Resolve` doesn't know what to do with a `YES` string.